internal/proxy/prometheus.go

```diff
@@ -0,0 +1,34 @@
+package proxy
+
+import (
+	"context"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strings"
+)
+
+type promURL string
+
+func NewPrometheus(prefix string, endpoint *url.URL) http.HandlerFunc {
+	var promKey promURL
+
+	director := func(r *http.Request) {
+		prometheusURL := r.Context().Value(promKey).(*url.URL)
+
+		r.URL.Scheme = prometheusURL.Scheme
+		r.URL.Host = prometheusURL.Host
+		// Normalize request so Prometheus doesn't know that request was proxied.
+		r.URL.Path = strings.TrimPrefix(r.URL.Path, prefix)
+	}
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		// Put the upstream PrometheusURL into the request's context for the director.
+		r = r.WithContext(context.WithValue(r.Context(), promKey, endpoint))
+
+		proxy := &httputil.ReverseProxy{
```

Comment:
I think this is anti-patter to allocate new reverse proxy on every request :thinking: At least this is not like I was using it all the time (: See: https://github.com/improbable-eng/kedge/blob/master/pkg/winch/http/proxy.go

Reply:
This is something I have copied over from one of our old iterations on an equivalent API.
I'll check the link you provided and improve these bits. 👍